Add notes on overwriting gufunc inputs to docs (#4957)
numba/numba: docs/source/user/vectorize.rst

   Use it to ensure the generated code does not fallback to
   :term:`object mode`.

.. _overwriting-input-values:

Overwriting input values
------------------------

In most cases, writing to inputs may also appear to work - however, this
behaviour cannot be relied on. Consider the following example function::

   @guvectorize([(float64[:], float64[:])], '()->()')
   def init_values(invals, outvals):
       invals[0] = 6.5
       outvals[0] = 4.2

Calling the `init_values` function with an array of `float64` type results in
visible changes to the input::

   >>> invals = np.zeros(shape=(3, 3), dtype=np.float64)
   >>> outvals = init_values(invals)
   >>> invals
   array([[6.5, 6.5, 6.5],
          [6.5, 6.5, 6.5],
          [6.5, 6.5, 6.5]])
   >>> outvals
   array([[4.2, 4.2, 4.2],
       [4.2, 4.2, 4.2],
       [4.2, 4.2, 4.2]])

> This works because Numpy passes a pointer to the input data directly into the

Reviewer: ```suggestion
This works because NumPy can pass the input data directly into the `init_values` function as the data `dtype` matches that of the declared argument.
```
Or something like ^. I think this needs to not refer to pointers and note that it's the `dtype` match that prevents the copy.

> This works because Numpy can pass the input data directly into the `init_values`

Reviewer: ```suggestion
This works because NumPy can pass the input data directly into the `init_values`
```

`init_values` function. However, it may also create and pass in a temporary
array, in which case changes to the input are lost. For example, this can occur

> when casting is required. To demonstrate this, we can  use an array of `float32`

Reviewer: ```suggestion
when casting is required. To demonstrate, we can  use an array of `float32`
```
